### PR TITLE
fix: always use DOMParser

### LINKED ISSUE #18: `createFragment` with `<img src="xxx">` will send HTTP request
### Initial checklist

- [X] I read the [support docs](https://github.com/rehypejs/.github/blob/main/support.md)
- [X] I read the [contributing guide](https://github.com/rehypejs/.github/blob/main/contributing.md)
- [X] I agree to follow the [code of conduct](https://github.com/rehypejs/.github/blob/main/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Arehypejs&type=Issues) and couldn’t find anything (or linked relevant results below)

### Problem

https://stackoverflow.com/questions/12747350/using-documentfragment-to-parse-html-without-sending-http-requests

### Solution

Use `DOMParser` and strip outside nodes

### Alternatives

N/A

### PR DESCRIPTION
**Pull request by JounQin**
fix #18

cc @wooorm 

### Initial checklist

*   [x] I read the support docs 
*   [x] I read the contributing guide 
*   [x] I agree to follow the code of conduct 
*   [x] I searched issues and couldn’t find anything (or linked relevant results below) 
*   [x] If applicable, I’ve added docs and tests

### Description of changes

N/A

### CODE REVIEW

**Comment by JounQin** [comment]
```log
/workspace/rehype-dom/packages/rehype-dom-parse/lib/index.js:39:7: child
/workspace/rehype-dom/packages/rehype-dom-parse/lib/index.js:41:11: child
46 / 48 95.83%
```

I'm not sure which case should I add?

**Comment by wooorm** [comment]
`child` is untyped. It needs to be typed. The previous way of first assigning, types it.

**Comment by wooorm** [comment]
This doesn’t work for elements that can be both in `head` and in `body`:

```js
rehypeDom()
  // Or `<!--comment-->`, or `<meta>`, or `<script>`, `<noscript>`, ...etc
  .process('<link rel="alternate" href="#">')
  .then((file) => {
    console.log(String(file))
  })
```

**Comment by wooorm** [comment]
This most likely results in the fragment being seen as quirks mode as well? ~~and perhaps even an unknown character set?~~

**Comment by JounQin** [comment]
> This doesn’t work for elements that can be both in head and in body:

Yeah, I just found that, now it's working except for `<!--comment-->`.

**Comment by JounQin** [comment]
It seems the simplest way is

```js
createDocument(`<body>${htmlString}</body>`)
```

File: packages/rehype-dom-parse/lib/index.js [hunk 1]
```diff
@@ -34,13 +34,13 @@ export default function parse(options) {
...
-  const temporary = document.createElement('body')
-  temporary.innerHTML = htmlString
-  let child = temporary.firstChild
+  const doc = createDocument(`<body>${htmlString}</body>`)
 
-  while (child) {
+  /** @type {Node | null} */
+  let child
+
+  while ((child = doc.body.firstChild)) {
```
Comment (wooorm) [inline packages/rehype-dom-parse/lib/index.js]:
```suggestion
  let child = doc.body.firstChild

  while (child) {
```

File: packages/rehype-dom-parse/lib/index.js [hunk 2]
```diff
@@ -34,13 +34,13 @@ export default function parse(options) {
...
     fragment.append(child)
```
Comment (wooorm) [inline packages/rehype-dom-parse/lib/index.js]:
```suggestion
    fragment.append(child)
    child = doc.body.firstChild
```

**Comment by wooorm** [comment]
Could you add a test for comments?

File: packages/rehype-dom-parse/lib/index.js (hunk 2 above)
Comment (JounQin) [reply packages/rehype-dom-parse/lib/index.js]: Fine, although I prefer the previous one. 🤣

**Comment by JounQin** [comment]
> Could you add a test for comments?

OK

**Comment by wooorm** [comment]
Could you also test locally whether an HTTP request might be kicked off?
I have a feeling that it might be the case if elements are moved from DOMParser to a fragment? 🤔

**Comment by JounQin** [comment]
> Could you also test locally whether an HTTP request might be kicked off?
> I have a feeling that it might be the case if elements are moved from DOMParser to a fragment? 🤔

Oh, it seems true. So, we should not use `DocumentFragment` at all...

**Comment by wooorm** [comment]
You can pass `document: true` if you want a document?

**Comment by JounQin** [comment]
> You can pass `document: true` if you want a document?

The problem is that I don't want a document...

I'll try to find a better way to fix this issue. `DocumentFragment` is actually hidden inside.

**Comment by wooorm** [comment]
I don't understand the problem:
- The goal of this project is to create a dom node that you render somewhere
- you are going to render that image, it just makes things faster if the request is dispatched as early as possible.
- You can do the same thing on your side?

**Comment by JounQin** [comment]
> you are going to render that image, it just makes things faster if the request is dispatched as early as possible.

I need to replace the `src` on rendering.

And `parse` should not do anything else IMO.

**Comment by JounQin** [comment]
```js
const DOCUMENT_FRAGMENT_NODE = 11

/**
 * @param {string} htmlString
 * @returns {Node}
 */
function createFragment(htmlString) {
  const doc = createDocument(`<body>${htmlString}</body>`)

  return /** @type {Node} */ ({
    nodeType: DOCUMENT_FRAGMENT_NODE,
    childNodes: doc.body.childNodes,
  })
}
```

How about this one?

**Comment by wooorm** [comment]
> I need to replace the src on rendering.

Do that on the server.

> And parse should not do anything else IMO.

Impossible in browsers. Or use `rehype-parse` itself.

> How about this one?

- Have you tested it? Does it work?
- You can do the same on your side
- It isn’t an actual fragment. The whole reason of the default is to create an actual fragment. Set `document: true` if you want a document

**Comment by JounQin** [comment]
> Impossible in browsers. Or use rehype-parse itself.

`DOMParser` does nothing else.

> Have you tested it? Does it work?

Yes, it's working.

> You can do the same on your side

I'm not sure how to.

> It isn’t an actual fragment.

It doesn't need to be. The user does not need to know the details inside.

**Comment by wooorm** [comment]
> `DOMParser` does nothing else.

Indeed, and `rehype-dom-parse`, in fragment mode, does nothing else than return a fragment. Set `document: true` if you want a document.
This is documented: https://github.com/rehypejs/rehype-dom/tree/main/packages/rehype-dom-parse#optionsfragment.

> > It isn’t an actual fragment.
>
> It doesn't need to be. The user does not need to know the details inside.

This project returns actual DOM nodes. Users expect actual DOM nodes and not similar objects.

---

> You can do the same on your side

Can you please share what you are working on, what you tried, what didn’t work?

**Comment by JounQin** [comment]
> This project returns actual DOM nodes. Users expect actual DOM nodes and not similar objects.

I'm not sure to understand, this project returns unified ast rather than actual DOM nodes, `createFragment` is only used inside, no user would expect actual DOM nodes, and it can't.

**Comment by wooorm** [comment]
Ahhh, I didn’t understand that, thanks!

File: packages/rehype-dom-parse/lib/index.js [hunk 3]
```diff
@@ -28,22 +28,23 @@ export default function parse(options) {
...
-  temporary.innerHTML = htmlString
-  let child = temporary.firstChild
-
-  while (child) {
-    fragment.append(child)
-    child = temporary.firstChild
-  }
-
-  return fragment
+  const doc = createDocument(`<body>${htmlString}</body>`)
```
Comment (wooorm) [inline packages/rehype-dom-parse/lib/index.js]:
```suggestion
  const doc = createDocument('<!doctype html><body>' + htmlString)
```

Just to make sure that it’s definitely seen as HTML5 and not as quirksmode or so.
And the `</body>` is unneeded here, so just the prefix does the same!

**Review by wooorm — COMMENTED** [review]
One last nit!

**Review by wooorm — APPROVED** [review]

**Comment by wooorm** [comment]
Thank Joun, released! ✨

**Pull request merged**